asm/metadata.go

```diff
@@ -0,0 +1,83 @@
+package asm
+
+// Metadata contains metadata about an instruction.
+type Metadata struct {
+	head *metaElement
+}
+
+type metaElement struct {
+	next       *metaElement
+	key, value interface{}
+}
+
+// Find the element containing key.
+//
+// Returns nil if there is no such element.
+func (m *Metadata) find(key interface{}) *metaElement {
+	for e := m.head; e != nil; e = e.next {
+		if e.key == key {
+			return e
+		}
+	}
+	return nil
+}
+
+// Remove an element from the linked list.
+//
+// Returns the new head of the list.
+func (m *Metadata) remove(r *metaElement) *metaElement {
+	var head *metaElement
+	prev := &head
+	for e := m.head; e != nil; e = e.next {
+		if e == r {
+			// We've found the element we want to remove. No need to copy the tail.
+			*prev = e.next
+			break
+		}
+
+		cpy := &metaElement{key: e.key, value: e.value}
+		*prev = cpy
+		prev = &cpy.next
```

Comment:
This line only seems to be hit once during the test suite, and the tests pass if it's removed. Could we add another test to exercise this better? (e.g. removing an element from a list where more than 1 element is copied?)

This is my understanding of how it works: after a copy, `prev` is supposed to point at the slot the next item will occupy, iow; it's supposed to point at a nil pointer initially. This fact is not 100% clear, but then I'm not super familiar with the data structure.

Nit: I think the name 'prev' is a bit misleading here, it's more like the 'current' item being operated on. My confusion during the first read-through was more due to lack of context, so just adding some comments about what each variable's meaning is would help a lot. :+1:

Reply:
Renamed to current.

Reply:
Also extended the test to have more bits of metadata.